clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -580,6 +581,11 @@ void EmitAssemblyHelper::CreatePasses(legacy::PassManager &MPM,
   if (CodeGenOpts.DisableLLVMPasses)
     return;
 
+  // Customize the head of the module passes list for the ESIMD extension.
+  if (LangOpts.SYCLIsDevice && LangOpts.SYCLExplicitSIMD) {
+    MPM.add(createSYCLLowerESIMDPass());
+  }
```

Comment:
Why don't you add it along with others that are needed for ESIMD?

Reply:
To run it first before all other opts. This is logically a part of FE rather than optimizer.